L15-PHP-Interceptors.md

```diff
@@ -0,0 +1,237 @@
+Title
+----
+* Author(s): ddyihai
+* Status: Draft
+* Implemented in: PHP
+
+## Abstract
+
+This proposal introduces the interceptor for gRPC PHP.
+__
+## Background
+
+The PHP language implementation for gRPC does not currently have an
+interceptor interface, whereas other languages like Go and Java do.
+Application concerns and extensibility are currently closed in the PHP
+implementation, making it difficult to extend the libraries for various
+application needs.
+
+
+### Related Proposals
+
+* [Python Client and Server Interceptors](https://github.com/grpc/proposal/pull/39)
+* [Ruby Client and Server Interceptors](https://github.com/grpc/proposal/pull/34)
+* [C# Client and Server Interceptors](https://github.com/grpc/proposal/pull/38)
+
+## Proposal
+
+Add client interceptors to the PHP gRPC implementation. Similar to the
+implementation in python, the channel is wrapper around an intercepted channel
+and the stub can use it the same way as the normal `Grpc\Channel`. RPCs dispatched 
+from such stubs then go through the interceptor chain for processing before they
+hit the underlying gRPC channel. 
+
+### Client Interceptors
+
+PHP provides a function `Grpc\Interceptor::intercept($channel, $interceptor)`
+to create a chain of interceptor, the API looks like below:
```

Comment:
> PHP provides a function `Grpc\Interceptor::intercept($channel, $interceptor)`
>  to create a chain of interceptor, the API looks like below:

Consider replacement with the following:

> In this proposal, intercepted channels are created using the new `Grpc\Interceptor::intercept($channel, $interceptor)` function.
This function takes a channel, and one or more interceptor objects and returns an intercepted channel. The return value is an object that shares the same public interface as the `Grpc\Channel` class and can be used in place of a `Grpc\Channel` object wherever that is accepted in the API, *e.g.* when instantiating a stub, or to pass it to the `Grpc\Interceptor::intercept` function.
>
> When a stub is created with an intercepted channel, the RPCs invoked through that stub will be intercepted by the interceptors registered on the channel that is used to create the stub.